interfaces/builtin/polkit.go

```diff
@@ -0,0 +1,156 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/polkit"
+	"github.com/snapcore/snapd/polkit/validate"
+	"github.com/snapcore/snapd/snap"
+)
+
+const polkitSummary = `allows access to polkitd to check authorisation`
+
+const polkitBaseDeclarationPlugs = `
+  polkit:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const polkitBaseDeclarationSlots = `
+  polkit:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const polkitConnectedPlugAppArmor = `
+# Description: Can talk to polkitd's CheckAuthorization API
+
+#include <abstractions/dbus-strict>
+
+dbus (send)
+    bus=system
+    path="/org/freedesktop/PolicyKit1/Authority"
+    interface="org.freedesktop.PolicyKit1.Authority"
+    member="{,Cancel}CheckAuthorization"
+    peer=(name="org.freedesktop.PolicyKit1", label=unconfined),
```

Comment:
Please forgive me if I'm misunderstanding the meaning of `slot-snap-type: [core]` above, but if that means that in Ubuntu Core this interface slots can be provided by a snap (I see that we have `implicitOnCore: false`), wouldn't that mean that this snap shipping the policy kit implementation would be running under a different AppArmor profile than `unconfined`? If so, then we don't want the `label=unconfined` condition here, do we?

Reply:
no, as defined here the interface works only on classic.  slot-snap-type: [core] means the slot can appear only on the system snap (core or snapd) and then that is controlled by the implicitOn* flags, this is a historical naming artifact when we didn't have yet the snapd snap. I have considered introducing "system" and deprecating "core" in this context but it would mean quite a bit on churn in some of the store snap-declarations.

Reply:
Aha!, that explains it. No worries then :-)